app/routes/session.py

```diff
@@ -169,9 +175,29 @@ def _set_questionnaire_supplementary_data(
         survey_id=metadata["survey_id"],
         sds_dataset_id=new_sds_dataset_id,
     )
+    # ensure any required lists for the schema are included in the supplementary data
+    _validate_supplementary_data_lists(
+        supplementary_data=supplementary_data, schema=schema
+    )
     questionnaire_store.set_supplementary_data(supplementary_data["data"])
 
 
+def _validate_supplementary_data_lists(
+    *, supplementary_data: dict, schema: QuestionnaireSchema
+) -> None:
+    """
+    Validates that any lists the schema requires (dynamic answers, list collector content blocks, repeating sections etc.) are covered:
+    either by a supplementary data list, or a list populated by a list collector
+    """
+    supplementary_lists = set(supplementary_data["data"].get("items", {}).keys())
+    populated_lists = supplementary_lists | schema.lists_populated_by_list_collector
+    if not schema.schema_dependent_lists <= populated_lists:
```

Comment:
I might be wrong here but I'm not following why we do this.
The schema will have a `supplementary_data.lists` property, so don't we just need to check `supplementary_data.lists` exists within `supplementary_lists`?. Everything else is checked by validator?